### PR TITLE
gpl: avoid crash by dividing to zero

### PR DESCRIPTION
**Pull request by gudeh**
This fix avoids a crash reported at issue #5197.

### CODE REVIEW

File: src/gpl/src/nesterovBase.cpp [hunk 1]
```diff
@@ -602,12 +602,21 @@ void BinGrid::initBins()
...
       = static_cast<int64_t>(ux_ - lx_) * static_cast<int64_t>(uy_ - ly_);
 
-  int64_t averagePlaceInstArea
-      = pb_->placeInstsArea() / pb_->placeInsts().size();
+  int64_t averagePlaceInstArea = 0;
+  if (!pb_->placeInsts().empty()) {
+    averagePlaceInstArea = pb_->placeInstsArea() / pb_->placeInsts().size();
+  }
+
+  int64_t idealBinArea = 0;
```
Comment (QuantamHD) [inline src/gpl/src/nesterovBase.cpp:610]: Is this a situation we want to recover from or send an error to the user? Separately, should placeInsts ever be empty at this point?

Comment (maliberty) [reply src/gpl/src/nesterovBase.cpp:610]: If there are no instances to place then we should skip all of gpl. Does this design really have no instances to place?

Comment (gudeh) [reply src/gpl/src/nesterovBase.cpp:610]: I do see instances placed at the beginning of GPL with debug mode. I will investigate further. Either way, I think the checking I have added so far does not hurt.

**Comment by gudeh** [comment]
As mentioned in the issue #5197, the reason we had a PlacerBase object without any placed instances is because the code creates a PlacerBase object for each group and another one for instances not related to any group (the first one in `pbVec_` on replace.cpp). Consequently, along the algorithm we would end up with division by zero and NANs.

To solve this issue I erased the first position of `pbVec_` when there are no placed instances related to it.

I ran a secure-CI and it passes (there are some failings matching master).

What do you guys think?

File: src/gpl/src/nesterovBase.cpp [hunk 2]
```diff
@@ -602,12 +602,23 @@ void BinGrid::initBins()
   int64_t totalBinArea
       = static_cast<int64_t>(ux_ - lx_) * static_cast<int64_t>(uy_ - ly_);
 
-  int64_t averagePlaceInstArea
-      = pb_->placeInstsArea() / pb_->placeInsts().size();
+  int64_t averagePlaceInstArea = 0;
+  if (!pb_->placeInsts().empty()) {
```
[the remark this answers is not in the document]
Comment (gudeh) [reply src/gpl/src/nesterovBase.cpp:606]: I am not sure. I understand this protection to avoid a crash is only beneficial, right?

**Review by maliberty — APPROVED** [review]

File: src/gpl/src/replace.cpp [hunk 3]
```diff
@@ -230,6 +230,10 @@ void Replace::doInitialPlace()
       }
     }
 
+    if (pbVec_.front()->placeInsts().empty()) {
+      pbVec_.erase(pbVec_.begin());
+    }
```
[the remark this answers is not in the document]
Comment (gudeh) [reply src/gpl/src/replace.cpp:235]: I tried defining a third region in the TCL and the PDN provided this error:
image

**Pull request merged**